### ISSUE TITLE
CI issues after the migration to the upstream emscripten backend

### ISSUE THREAD

**Issue by rth** [iodide-project/pyodide#1204]
After the migration to the upstream emscripten backend was merged https://github.com/iodide-project/pyodide/pull/1102, there have been quite a few CI failure on master (cf below).  The CI in the PR was green aside for one timeout in chrome packages.

Currently following issues have been observed in the CI on master (even after restarting the CI once),
 - [x] firefox core:  failures in webworker tests,
      ```
      =========================== short test summary info ============================
      FAILED src/tests/test_webworker.py::test_runwebworker_different_package_name[firefox]
      FAILED src/tests/test_webworker.py::test_runwebworker_no_imports[firefox] - s...
      FAILED src/tests/test_webworker.py::test_runwebworker_missing_import[firefox]
      FAILED src/tests/test_webworker.py::test_runwebworker_exception[firefox] - se...
      FAILED src/tests/test_webworker.py::test_runwebworker_exception_after_import[firefox]
      FAILED src/tests/test_webworker.py::test_runwebworker_micropip[firefox] - sel...
      ===== 6 failed, 1 passed, 1 skipped, 1136 deselected in 313.01s (0:05:13) ======
      ```
 - [x] chrome and firefox core: occasional failure in test_keyboard_interrupt https://github.com/iodide-project/pyodide/issues/1200
 - [x] chrome core: occasional timeouts
 - [x] firefox and firefox packages: sometimes,
       ```
       packages/numpy/test_numpy.py::test_runwebworker_numpy[firefox] FAILED    [ 66%]
       packages/pandas/test_pandas.py::test_load_largish_file[firefox] FAILED   [100%]
       ```
       due to timeout other times other times pillow #1201. In chrome sometimes almost all package import fail (after 2hours because of a 30s timeout per test)

At the same time I think this is a CI issue because **locally code from master builds and tests fine** in the same Docker image (aside from test_keyboard_interrupt and test_pillow for me). Maybe we have to re-evaluate running some tests in parallel, and revert xfail scipy related tests as we did before https://github.com/iodide-project/pyodide/pull/1102 There are also other things we could explore.

Because updating to latest emscripten is critical, I think it might be good to try to fix it in the current state. Rather than revert and try to fix in a separate PR. Though this means that master will sometimes show a broken CI which will be an inconvinience for other PRs untill we fix it. If we can't find workarounds soon then we can revert (and re-open a new PR with fixes on top). WDYT?

cc @dalcde @hoodmane  @joemarshall

**Comment by hoodmane** [comment]
> Because updating to latest emscripten is critical, I think it might be good to try to fix it in the current state. Rather than revert and try to fix in a separate PR. Though this means that master will sometimes show a broken CI which will be an inconvinience for other PRs untill we fix it. If we can't find workarounds soon then we can revert. WDYT?

Sounds good to me. Not like the CI worked great even before.

**Comment by rth** [comment]
I also just realized that the selenium version we are using is from Nov 2018, and there have been no stable releases since (though there are fairly recent alpha pre-releases 4.0.0.a7). So that also probably doesn't help. Will investigate their alpha release.

**Comment by casatir** [comment]
As a side note,

- our `Firefox` version (70.0.1) is from Oct 2019
- our `geckodriver` version (0.26.0), from Oct 2019
- our `Chromium` version (87.0.4280.88) is from Dec 2020
- and our `ChromeDriver` version (2.41.578700) from Aug 2018 (and is advertised for Chrome v67-69 !)

**Comment by rth** [comment]
Yes, that's another direction to explore. For firefox https://github.com/iodide-project/pyodide/issues/716 was a blocker at the time. ChromeDriver clearly needs updating.

**Comment by rth** [comment]
After connecting to a CircleCI instance, it appears that CPU resources are badly over-subscribed,

![image](https://user-images.githubusercontent.com/630936/107151814-ec7a0d80-6964-11eb-9cb7-5cbbc7ab089d.png)

260 load average for 36 core (or maybe /2 with multi-threading). We get 2 vCPU, I'm not sure what exactly this means and how access to CPU resources is restricted in Docker in their setup. 
When staring for instance 3 parallel jobs with pytest-xdist, this appears to create 3 process for each pytest-xdist job (x3) with with ~20 threads each.

![image](https://user-images.githubusercontent.com/630936/107151969-bd17d080-6965-11eb-8af8-faed9e71fc57.png)

Part of it might be that anything that auto-detects the number of available CPUs will detect 36 instead of 2 we are allowed to use, hence the CPU oversubscription even for our own processes.

**Comment by rth** [comment]
Continuing the investigation of webworker related failures on Firefox. For the test that just returns a number from a webworker,
```
pytest src/tests/test_webworker.py -v -k "firefox and test_runwebworker_no_imports" -x --durations 10
```
below is the timing on a desktop (test passes)
```
2.89s setup    src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
1.60s call     src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
0.51s teardown src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
```
and the timing in CircleCI (test fails),
```
30.05s call     src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
17.70s setup    src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
1.93s teardown src/tests/test_webworker.py::test_runwebworker_no_imports[firefox]
```
So the selenium setup is 6x slower in CI, however even increasing the script timeout from 30s to 5min,
```
        self.driver.set_script_timeout(60*3)
```
it still consistently gets timeouts. And then again on other CI runs this test passes. So I have the impression that the issue is a mis-configuration in our testing setup for webworkers that leads to deadlocks.

**Comment by rth** [comment]
OK, in the end it looks like all of this was a CircleCI service issue https://status.circleci.com/incidents/qf46sjvhzvlz and CI is back to green aside from an occasional chrome packages timeout. I guess hard to believe coincidences do happen )
